File: src/frontend/lib/btw.ts
/**
 * BTW (VAT) aangifte utilities.
 * All amounts in integer euro cents internally.
 */

export interface BtwAangifteResponse {
  id: string;
  year: number;
  quarter: number;
  status: "draft" | "submitted" | "accepted";
  box_1a_net_cents: number;
  box_1b_net_cents: number;
  box_1c_net_cents: number;
  box_1d_net_cents: number;
  box_5a_vat_due_cents: number;
  box_5b_voorbelasting_cents: number;
  box_5d_payable_cents: number;
  notes: string | null;
  submitted_at: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * Format euro cents to Dutch locale string: "€ 1.234,56"
 */
export function formatBtwCents(cents: number): string {
  const euros = cents / 100;
  return (
    "€ " +
    euros.toLocaleString("nl-NL", {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    })
  );
}

const QUARTER_LABELS: Record<number, string> = {
  1: "jan-mrt",
  2: "apr-jun",
  3: "jul-sep",
  4: "okt-dec",
};

/**
 * Format a quarter label in Dutch: "Q1 2024 (jan-mrt)"
 */
export function formatQuarterLabel(year: number, quarter: number): string {
  return `Q${quarter} ${year} (${QUARTER_LABELS[quarter]})`;
}

/**
 * Return ISO date range strings for a quarter.
 */
export function getQuarterDateRange(
  year: number,
  quarter: number
): { start: string; end: string } {
  const firstMonth = (quarter - 1) * 3 + 1;
  const lastMonth = firstMonth + 2;
  const lastDay = new Date(year, lastMonth, 0).getDate();
  const pad = (n: number) => String(n).padStart(2, "0");
  return {
    start: `${year}-${pad(firstMonth)}-01`,
    end: `${year}-${pad(lastMonth)}-${pad(lastDay)}`,
  };
}

File: src/backend/app/schemas/btw.py
"""Pydantic schemas for BTW aangifte endpoints."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

BtwStatus = Literal["draft", "submitted", "accepted"]


class BtwGenerateRequest(BaseModel):
    year: int = Field(ge=2000, le=2100)
    quarter: int = Field(ge=1, le=4)


class BtwAangifteUpdate(BaseModel):
    notes: str | None = None
    status: BtwStatus | None = None
    # Override fields (manual corrections by accountant)
    box_1a_net_cents: int | None = Field(default=None, ge=0)
    box_1b_net_cents: int | None = Field(default=None, ge=0)
    box_1c_net_cents: int | None = Field(default=None, ge=0)
    box_1d_net_cents: int | None = Field(default=None, ge=0)
    box_5a_vat_due_cents: int | None = Field(default=None, ge=0)
    box_5b_voorbelasting_cents: int | None = Field(default=None, ge=0)
    box_5d_payable_cents: int | None = None


class BtwAangifteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    year: int
    quarter: int
    status: str

    box_1a_net_cents: int
    box_1b_net_cents: int
    box_1c_net_cents: int
    box_1d_net_cents: int
    box_5a_vat_due_cents: int
    box_5b_voorbelasting_cents: int
    box_5d_payable_cents: int

    notes: str | None
    submitted_at: datetime | None
    created_at: datetime
    updated_at: datetime
